File: packages/feedback/src/threshold.ts
import { createFeedbackResolver } from "@composable-model-graph/core";
import type { FeedbackResolver } from "@composable-model-graph/core";

/** Options for {@link thresholdFeedbackResolver}. */
export interface ThresholdFeedbackResolverOptions {
  id?: string;
  name?: string;
  /** Score at or above which the run is accepted. */
  acceptScore: number;
  /**
   * Optional score below which the run should be retried. Scores between
   * `retryScore` and `acceptScore` map to `adjust`. When omitted, anything
   * below `acceptScore` maps to `adjust`.
   */
  retryScore?: number;
}

/**
 * Decide a feedback action from the evaluation score.
 *
 *   score >= acceptScore                 -> accept
 *   retryScore set and score < retryScore -> retry
 *   otherwise (score present)             -> adjust
 *   score absent                          -> custom (inspect)
 */
export function thresholdFeedbackResolver<I = unknown, O = unknown>(
  options: ThresholdFeedbackResolverOptions,
): FeedbackResolver<I, O> {
  const { acceptScore, retryScore } = options;

  return createFeedbackResolver<I, O>({
    id: options.id ?? "threshold-feedback",
    name: options.name ?? "Threshold Feedback",
    resolve: (run) => {
      const score = run.evaluation?.score;
      if (score === undefined) {
        return {
          kind: "custom",
          reason: "no score to threshold",
          signal: { action: "inspect" },
        };
      }
      if (score >= acceptScore) {
        return { kind: "accept", reason: `score ${score} >= ${acceptScore}` };
      }
      if (retryScore !== undefined && score < retryScore) {
        return { kind: "retry", reason: `score ${score} < ${retryScore}` };
      }
      return { kind: "adjust", reason: `score ${score} < ${acceptScore}` };
    },
  });
}

File: python/src/composable_model_graph/feedback/threshold.py
"""Decide a feedback action from the evaluation score.

    score >= accept_score                  -> accept
    retry_score set and score < retry_score -> retry
    otherwise (score present)               -> adjust
    score absent                            -> custom (inspect)

(Python parity of typescript feedback/threshold.ts.)
"""

from __future__ import annotations

from typing import Any, Optional

from ..core import FeedbackAction, FeedbackResolver, create_feedback_resolver
from ..core._format import format_number as _fmt


def threshold_feedback_resolver(
    accept_score: float,
    retry_score: Optional[float] = None,
    id: Optional[str] = None,
    name: Optional[str] = None,
) -> FeedbackResolver:
    def resolve(run: Any, context: Any) -> FeedbackAction:
        score = run.evaluation.score if run.evaluation else None
        if score is None:
            return FeedbackAction(
                kind="custom",
                reason="no score to threshold",
                signal={"action": "inspect"},
            )
        if score >= accept_score:
            return FeedbackAction(
                kind="accept", reason=f"score {_fmt(score)} >= {_fmt(accept_score)}"
            )
        if retry_score is not None and score < retry_score:
            return FeedbackAction(
                kind="retry", reason=f"score {_fmt(score)} < {_fmt(retry_score)}"
            )
        return FeedbackAction(
            kind="adjust", reason=f"score {_fmt(score)} < {_fmt(accept_score)}"
        )

    return create_feedback_resolver(
        id=id if id is not None else "threshold-feedback",
        name=name if name is not None else "Threshold Feedback",
        resolve=resolve,
    )
